src/sqlfluff/core/parser/grammar/base.py

```diff
@@ -343,54 +343,40 @@ def _trim_elem(seg):
                 s = seg.raw_upper.split(maxsplit=1)
                 return s[0] if s else ""
 
-            str_buff = [_trim_elem(seg) for seg in segments]
-            match_queue = []
-
-            for matcher, simple in simple_matchers:
-                # Simple will be a tuple of options
-                assert simple
-                for simple_option in simple:
-                    # NOTE: We use iter_indices to make sure we capture
-                    # all instances of potential matches if there are many.
-                    # This is important for bracket counting.
-                    for buff_pos in iter_indices(str_buff, simple_option):
-                        match_queue.append((matcher, buff_pos, simple_option))
-
-            match_queue = sorted(match_queue, key=lambda x: x[1])
-
             parse_match_logging(
                 cls.__name__,
                 "_look_ahead_match",
                 "SI",
                 parse_context=parse_context,
                 v_level=4,
-                mq=match_queue,
-                sb=str_buff,
             )
 
-            while match_queue:
+            simple_match = None
+            for idx, seg in enumerate(segments):
+                for matcher, simple in simple_matchers:
+                    # Simple will be a tuple of options
+                    assert simple
+                    for simple_option in simple:
+                        if simple_option == _trim_elem(seg):
```

Comment:
I think we could further optimise this by computing `_trim_elem(seg)` outside this inner `for` loop, because it's always the same for a given `seg`.

Reply:
Maybe even better would be caching the trimmed element on the segment so that it's just a cache fetch.